cmd/proxy-server/main.go

```diff
@@ -540,3 +562,157 @@ func rolesHandler(log *logrus.Entry, opaHost string) http.Handler {
+			}
+
+			log.Debugf("Serving get volumes request for tenant %s", claims.Group)
+
+			if roleServ.roleService == nil {
+				resp, err = roleServ.roleClient.List(r.Context(), &pb.RoleListRequest{})
+			} else {
+				resp, err = roleServ.roleService.List(r.Context(), &pb.RoleListRequest{})
+			}
+
+			if err != nil {
+				log.WithError(err).Printf("error get roles: %v", err)
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("error creating json response")
+				}
+				return
+			}
+
+			roleJSON := roles.NewJSON()
+			err = roleJSON.UnmarshalJSON(resp.Roles)
+			if err != nil {
+				log.WithError(err).Printf("error unmarshalling JSON: %v", err)
+				w.WriteHeader(http.StatusInternalServerError)
+				if err := web.JSONErrorResponse(w, err); err != nil {
+					log.WithError(err).Println("error creating json response")
+				}
+				return
+			}
+
+			rolesSplit := strings.Split(claims.Roles, ",")
+
+			roleJSON.Select(func(rInst roles.Instance) {
+				for _, role := range rolesSplit {
+					if rInst.Name == role {
+						log.Info("found role!")
+						sysID = rInst.SystemID
+						storPool = rInst.Pool
+						sysType = rInst.SystemType
+						tenant = claims.Group
+
+						dataKey := fmt.Sprintf("quota:%s:%s:%s:%s:data", sysType, sysID, storPool, tenant)
+
+						res, err := rdb.HGetAll(dataKey).Result()
+						if err != nil || len(res) == 0 {
+							log.WithError(err).Printf("no volumes found for tenant %s, %v", tenant, err)
+							w.WriteHeader(http.StatusInternalServerError)
+							if err := web.JSONErrorResponse(w, fmt.Errorf("no volumes found for tenant %s", tenant)); err != nil {
+								log.WithError(err).Println("error creating json response")
+							}
+							return
+						}
+
+						for volKey := range res {
+							if strings.Contains(volKey, "capacity") {
+								splitStr := strings.Split(volKey, ":")
+								//example : vol:k8s-cb89d36285:capacity
+								if len(splitStr) == 3 {
+									volumeMap[splitStr[1]] = splitStr[1]
+								}
+							}
+							if strings.Contains(volKey, "deleted") {
+								splitStr := strings.Split(volKey, ":")
+								//example : vol:k8s-cb89d36285:deleted
+								if len(splitStr) == 3 {
+									delete(volumeMap, splitStr[1])
+								}
+							}
+						}
+					}
+				}
+			})
+
+		case "Basic":
+			log.Println("Basic authentication used")
+			return
+		}
+		if len(volumeMap) == 0 {
+			log.Errorf("no volumes found for tenant %s", tenant)
+			w.WriteHeader(http.StatusInternalServerError)
+			if err := web.JSONErrorResponse(w, fmt.Errorf("no volumes found for tenant %s", tenant)); err != nil {
+				log.WithError(err).Println("error creating json response")
+			}
+		}
+
+		//append resulting map values onto volumeList
+		for _, v := range volumeMap {
+			volumeList = append(volumeList, v)
+		}
+
+		log.Printf("volumeList %+v\n", volumeList)
+		//TODO: grpc call to storage service to get volume details
```

Comment:
should this TODO be removed?

Reply:
Nopes, it will get removed after integration